protocols/gossipsub/src/behaviour.rs

```diff
@@ -921,16 +963,63 @@ impl Gossipsub {
 
             for peer in recipient_peers.iter() {
                 debug!("Sending message: {:?} to peer {:?}", msg_id, peer);
-                self.events.push_back(NetworkBehaviourAction::NotifyHandler {
-                    peer_id: peer.clone(),
-                    event: event.clone(),
-                    handler: NotifyHandler::Any,
-                });
+                self.events
+                    .push_back(NetworkBehaviourAction::NotifyHandler {
+                        peer_id: peer.clone(),
+                        event: event.clone(),
+                        handler: NotifyHandler::Any,
+                    });
             }
         }
         debug!("Completed forwarding message");
     }
 
+    /// Constructs a `GossipsubMessage` performing message signing if required.
+    pub(crate) fn build_message(
+        &self,
+        topics: Vec<TopicHash>,
+        data: Vec<u8>,
+    ) -> Result<GossipsubMessage, String> {
+        let sequence_number: u64 = rand::random();
+        let message = rpc_proto::Message {
+            from: Some(self.message_source_id.clone().into_bytes()),
+            data: Some(data),
+            seqno: Some(sequence_number.to_be_bytes().to_vec()),
+            topic_ids: topics.clone().into_iter().map(|t| t.into()).collect(),
+            signature: None,
+            key: None,
+        };
+
+        // If a signature is required, generate it
+        let signature = if let Some(keypair) = self.keypair.as_ref() {
```

Comment:
`message` is only needed when `Some(keypair) == self.keypair.as_ref()` right?

```suggestion
        // If a signature is required, generate it
        let signature = if let Some(keypair) = self.keypair.as_ref() {
              let message = rpc_proto::Message {
                from: Some(self.message_source_id.clone().into_bytes()),
                data: Some(data),
                seqno: Some(sequence_number.to_be_bytes().to_vec()),
                topic_ids: topics.clone().into_iter().map(|t| t.into()).collect(),
                signature: None,
                key: None,
            };
```

Reply:
This is true, but if i recall correctly, I was trying to avoid the clone on data. 
I'll change it, but it now clones the data.

Reply:
I see. I guess as this is only scoped to the function itself we can delay the decision. In case it shows up in benchmarks we can easily change it. Rustc or LLVM might as well just optimize it away, not sure.